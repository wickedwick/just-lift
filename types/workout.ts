export type Exercise = {
  id: string,
  name: string
  sets: number
  reps: number
  weight: number
  weightUnit: WeightUnit
  progressiveOverload: boolean
  overloadIncrement: number
}

export type Workout = {
  id: string
  exercises: Exercise[]
}

export type WorkoutPlan = {
  workouts: Workout[]
  daysPerWeek: number
  workoutIndex: number
  workoutInProgress: boolean
}

export type Log = {
  date: Date
  workoutId: string
  data: string
  exerciseName: string
  weight: number
  weightUnit: WeightUnit
}

export enum WeightUnit {
  Lbs,
  Kg
}
 