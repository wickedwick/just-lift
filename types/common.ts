import { Exercise, Log, WorkoutPlan } from './workout';
import { StyleProp, TextStyle, ViewStyle } from 'react-native';

export type RootStackParamList = {
  Root: undefined,
  NotFound: undefined
}

export type BottomTabParamList = {
  AutoTagger: undefined
  Celebrities: undefined
}

export type TabOneParamList = {
  BlankMenuScreen: undefined
  CreateWorkoutScreen: undefined
  SelectWorkoutScreen: undefined
  WorkoutScreen: undefined
  SettingsScreen: undefined
  ProfileScreen: undefined
  GuidedSelectionScreen: undefined
  WorkoutListScreen: undefined
}

export type TabTwoParamList = {
  TabOneScreen: undefined
  TabTwoScreen: undefined
}

export type WorkoutFormProps = {
  exercise?: Exercise
  onSubmit: (exercise: Exercise) => void
  onCancel: () => void
}

export type ExerciseCardProps = {
  exercise: Exercise
  onRemovePress: () => void
  onEditPress: () => void
}

export type ExerciseCounterProps = {
  exercise: Exercise
  setLogData: (log: Log) => void
  setWeight: (weight: number, exerciseName: string) => void
}

export type LogCardProps = {
  log: Log
}

export type ActionButtonProps = {
  onPress: () => void
  contained?: boolean
  icon?: string
  labelStyle?: StyleProp<TextStyle>
  style?: StyleProp<ViewStyle>
  text: string
}

export type RadioButtonProps = {
  label: string
  value: string
}

export type CheckboxProps = {
  label: string
  status: 'checked' | 'unchecked'
  onPress: () => void
}

export type WorkoutPlanContextProps = {
  workoutPlan: WorkoutPlan | null
  setWorkoutPlan: (workoutPlan: WorkoutPlan | null) => void
}
