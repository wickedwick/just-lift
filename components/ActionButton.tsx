import React from 'react';
import { ActionButtonProps } from '../types/common';
import { Button } from 'react-native-paper';

const ActionButton = (props: ActionButtonProps): JSX.Element => {
  return (
    <Button
      icon={props.icon}
      mode={props.contained ? 'contained' : 'text'}
      onPress={props.onPress}
      style={props.style}
      labelStyle={props.labelStyle}
    >
      {props.text}
    </Button>
  )
}

export default ActionButton
