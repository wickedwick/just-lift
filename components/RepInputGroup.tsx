import ActionButton from './ActionButton'
import React, { useState } from 'react'
import { Button, TextInput } from 'react-native-paper'
import { createGroups, getNumberOrDefault } from '../services/utils'
import { createLog } from '../services/log'
import { Log } from '../types/workout'
import { RepInputGroupProps } from '../types/common'
import { StyleSheet, TouchableOpacity, View } from 'react-native'

const RepInputGroup = (props: RepInputGroupProps): JSX.Element => {
  const { counts, exercise, setCounts, setLogData } = props
  const [repsEditingIndex, setRepsEditingIndex] = useState<number>(-1)

  const onChange = (index: number) => {
    let newCounts = [...counts]

    if (newCounts[index] > 0) {
      newCounts[index]--
    } else {
      newCounts[index] = exercise.reps
    }

    setCounts(newCounts)
    onCreateLog(newCounts)
  }

  const onRepsChange = (text: string, index: number) => {
    const newReps = getNumberOrDefault(text)
    const newCounts = [...counts]
    newCounts[index] = newReps
    setCounts(newCounts)
  }

  const onRepsEditingChange = () => {
    onCreateLog(counts)
    setRepsEditingIndex(-1)
  }

  const onCreateLog = (newCounts: number[]) => {
    const newLog: Log = createLog(exercise, newCounts)
    setLogData(newLog)
  }

  return (
    <View style={{ flex: 1, flexDirection: 'column' }}>
      {createGroups([...Array(exercise.sets).fill(0)], 7)
        .map((group: number[], outerIndex) => {
          return (
            <View style={styles.flexContainer} key={outerIndex}>
            {group.map((_, index) => (
              <TouchableOpacity
                activeOpacity={0.6}
                key={index}
                onLongPress={() => { setRepsEditingIndex(index) }}
              >
                {repsEditingIndex === (outerIndex * 7) + index ? (
                  <>
                    <TextInput
                      keyboardType={'numeric'}
                      label={`Reps`}
                      onChangeText={(text) => { onRepsChange(text, (outerIndex * 7) + index) }}
                      style={{ ...styles.textInput, marginBottom: 0 }}
                      value={counts[(outerIndex * 7) + index]?.toString()}
                    />
                    <ActionButton
                      contained
                      icon="check"
                      labelStyle={{ fontSize: 20, textAlign: 'center' }}
                      onPress={() => { onRepsEditingChange() }}
                      style={{ alignSelf: 'center', padding: 3, marginTop: 0 }}
                      text=""
                    />
                  </>
                ) : (
                  <Button
                    compact
                    key={index}
                    labelStyle={styles.buttonLabel}
                    mode='contained'
                    onPress={() => onChange((outerIndex * 7) + index)}
                    onLongPress={() => { setRepsEditingIndex((outerIndex * 7) + index) }}
                    style={styles.input}
                  >
                    {counts[(outerIndex * 7) + index]}
                  </Button>
                )}
              </TouchableOpacity>)
            )}
            </View>
          )
        }
      )}
    </View>
  )
}

const styles = StyleSheet.create({
  buttonLabel: {
    fontSize: 20,
    fontWeight: 'bold',
    alignContent: 'center',
    margin: 0,
  },
  flexContainer: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  input: {
    marginBottom: 10,
    marginTop: 10,
    width: 50,
    height: 50,
    borderRadius: 50,
    padding: 3,
  },
  textInput: {
    marginBottom: 10,
    marginTop: 10,
    padding: 3,
  },
})

export default RepInputGroup
