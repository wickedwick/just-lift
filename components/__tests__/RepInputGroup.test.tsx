import ActionButton from '../ActionButton'
import Adapter from 'enzyme-adapter-react-16'
import React from 'react'
import RepInputGroup from '../RepInputGroup'
import { configure, shallow } from 'enzyme'
import { RepInputGroupProps } from '../../types/common'
import { WeightUnit } from '../../types/workout'
import { TouchableOpacity } from 'react-native'
import { Button, TextInput } from 'react-native-paper'

configure({ adapter: new Adapter() })

const props: RepInputGroupProps = {
  counts: [0, 0, 0, 0],
  exercise: {
    name: 'Bench Press',
    progressiveOverload: false,
    overloadIncrement: 0,
    reps: 39,
    sets: 4,
    weight: 100,
    weightUnit: WeightUnit.Lbs,
  },
  setCounts: jest.fn(),
  setLogData: jest.fn(),
}

afterEach(() => {
  jest.clearAllMocks();
});

describe('<RepInputGroup />', () => {
  it('Renders', () => {
    const wrapper = shallow(<RepInputGroup {...props} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('Handles long press', () => {
    const wrapper = shallow(<RepInputGroup {...props} />)

    wrapper.find(TouchableOpacity).first().simulate('longPress')

    wrapper.setProps({ repsEditingIndex: 0 })
    wrapper.dive().find(TextInput).simulate('changeText', '68')
    expect(props.setCounts).toHaveBeenCalled()

    wrapper.find(ActionButton).simulate('press')
    expect(props.setLogData).toHaveBeenCalledTimes(1)
  })

  it('Handles button press', () => {
    const wrapper = shallow(<RepInputGroup {...props} />)

    wrapper.find(Button).first().simulate('press')
    expect(props.setCounts).toHaveBeenCalledTimes(1)
  })
})
