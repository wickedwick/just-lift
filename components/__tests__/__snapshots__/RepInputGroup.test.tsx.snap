// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RepInputGroup /> Renders 1`] = `ShallowWrapper {}`;
